In [84]:
import pandas as pd

In [85]:
df = pd.read_csv("weatherAUS.csv", parse_dates = ["Date"])
df["Year"] = df["Date"].dt.year
df["Month"] = df['Date'].dt.month_name()
df[["Location","WindGustDir","WindDir9am","WindDir3pm","RainToday","RainTomorrow","Month"]] = df[["Location","WindGustDir","WindDir9am","WindDir3pm","RainToday","RainTomorrow","Month"]].astype('category')
df.head(5)

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0 2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1 2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2 2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3 2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4 2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Pressure9am  Pressure3pm  \
0           W           44.0          W  ...      1007.7       1007.1   
1         WNW           44.0        NNW  ...      1010.6       1007.8   
2         WSW           46.0          W  ...      1007.6       1008.7   
3          NE           24.0         SE  ...      1017.6       1012.8   
4           W           41.0        ENE  ...      1010.8       1006.0   

   Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  RainTomorrow  Year  \
0       8.0       NaN     16.9     21.8         No            No  2008   
1       NaN       NaN     17.2     24.3         No            No  2008   
2       NaN       2.0     21.0     23.2         No            No  2008   
3       NaN       NaN     18.1     26.5         No            No  2008   
4       7.0       8.0     17.8     29.7         No            No  2008   

      Month  
0  December  
1  December  
2  December  
3  December  
4  December  

[5 rows x 25 columns]

In [86]:
df.nunique()

Date             3436
Location           49
MinTemp           389
MaxTemp           505
Rainfall          681
Evaporation       358
Sunshine          145
WindGustDir        16
WindGustSpeed      67
WindDir9am         16
WindDir3pm         16
WindSpeed9am       43
WindSpeed3pm       44
Humidity9am       101
Humidity3pm       101
Pressure9am       546
Pressure3pm       549
Cloud9am           10
Cloud3pm           10
Temp9am           441
Temp3pm           502
RainToday           2
RainTomorrow        2
Year               11
Month              12
dtype: int64

In [87]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 25 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   Date           145460 non-null  datetime64[ns]
 1   Location       145460 non-null  category      
 2   MinTemp        143975 non-null  float64       
 3   MaxTemp        144199 non-null  float64       
 4   Rainfall       142199 non-null  float64       
 5   Evaporation    82670 non-null   float64       
 6   Sunshine       75625 non-null   float64       
 7   WindGustDir    135134 non-null  category      
 8   WindGustSpeed  135197 non-null  float64       
 9   WindDir9am     134894 non-null  category      
 10  WindDir3pm     141232 non-null  category      
 11  WindSpeed9am   143693 non-null  float64       
 12  WindSpeed3pm   142398 non-null  float64       
 13  Humidity9am    142806 non-null  float64       
 14  Humidity3pm    140953 non-null  float64       
 15  

In [88]:
df.describe()

MinTemp        MaxTemp       Rainfall   Evaporation  \
count  143975.000000  144199.000000  142199.000000  82670.000000   
mean       12.194034      23.221348       2.360918      5.468232   
std         6.398495       7.119049       8.478060      4.193704   
min        -8.500000      -4.800000       0.000000      0.000000   
25%         7.600000      17.900000       0.000000      2.600000   
50%        12.000000      22.600000       0.000000      4.800000   
75%        16.900000      28.200000       0.800000      7.400000   
max        33.900000      48.100000     371.000000    145.000000   

           Sunshine  WindGustSpeed   WindSpeed9am   WindSpeed3pm  \
count  75625.000000  135197.000000  143693.000000  142398.000000   
mean       7.611178      40.035230      14.043426      18.662657   
std        3.785483      13.607062       8.915375       8.809800   
min        0.000000       6.000000       0.000000       0.000000   
25%        4.800000      31.000000       7.000000      13.000000   
50%        8.400000      39.000000      13.000000      19.000000   
75%       10.600000      48.000000      19.000000      24.000000   
max       14.500000     135.000000     130.000000      87.000000   

         Humidity9am    Humidity3pm   Pressure9am    Pressure3pm  \
count  142806.000000  140953.000000  130395.00000  130432.000000   
mean       68.880831      51.539116    1017.64994    1015.255889   
std        19.029164      20.795902       7.10653       7.037414   
min         0.000000       0.000000     980.50000     977.100000   
25%        57.000000      37.000000    1012.90000    1010.400000   
50%        70.000000      52.000000    1017.60000    1015.200000   
75%        83.000000      66.000000    1022.40000    1020.000000   
max       100.000000     100.000000    1041.00000    1039.600000   

           Cloud9am      Cloud3pm        Temp9am       Temp3pm           Year  
count  89572.000000  86102.000000  143693.000000  141851.00000  145460.000000  
mean       4.447461      4.509930      16.990631      21.68339    2012.769751  
std        2.887159      2.720357       6.488753       6.93665       2.537684  
min        0.000000      0.000000      -7.200000      -5.40000    2007.000000  
25%        1.000000      2.000000      12.300000      16.60000    2011.000000  
50%        5.000000      5.000000      16.700000      21.10000    2013.000000  
75%        7.000000      7.000000      21.600000      26.40000    2015.000000  
max        9.000000      9.000000      40.200000      46.70000    2017.000000

# Anlysis of whole Years

In [89]:
df[df["MinTemp"] == -8.5000] #detail when the Temp was lowest.

Date     Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
52254 2009-06-11  MountGinini     -8.5      0.6       1.0          NaN   

       Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Pressure9am  \
52254       NaN           S           31.0        WNW  ...         NaN   

       Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
52254          NaN       NaN       NaN     -0.9     -0.4         No   

       RainTomorrow  Year  Month  
52254            No  2009   June  

[1 rows x 25 columns]

In [90]:
df[df["MaxTemp"] == 48.1000] # details when the Temp was Highest

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
106316 2011-01-25  Woomera     25.1     48.1       0.0         22.4      12.8   

       WindGustDir  WindGustSpeed WindDir9am  ... Pressure9am  Pressure3pm  \
106316         WSW           85.0        NNE  ...      1003.4       1000.4   

        Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  RainTomorrow  Year  \
106316       3.0       2.0     33.2     46.1         No            No  2011   

          Month  
106316  January  

[1 rows x 25 columns]

In [91]:
df[df["Rainfall"]==371]

Date      Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
9368 2009-11-07  CoffsHarbour     17.4     23.1     371.0          NaN   

      Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Pressure9am  \
9368       3.5         NaN            NaN         SW  ...      1026.6   

      Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
9368       1025.9       8.0       5.0     18.6     22.0        Yes   

      RainTomorrow  Year     Month  
9368           Yes  2009  November  

[1 rows x 25 columns]

## Insides:
1. The lowest temperature was observed in MountGinni of magnitude -8.5 degree cel in June,2009.
2. The highest temperature was observed in Woomera of 48.1 degree cel in January,2011
3. Average, minimum and maximum Rainfall was 2.3609, 0.00 and 371.00 respectively.
4. Maximum Rainfall was 371.00 mm in Caffsharbour in November,2009.

In [92]:
pd.set_option("display.max_rows",None)
Detail = df.pivot_table(values = ['MinTemp',"MaxTemp","Rainfall"],index = ['Location','Year'], aggfunc = {"MinTemp":["min","mean"],"MaxTemp":["max","mean"],"Rainfall":["min","mean","max"]})
Detail["Diff of Temp"] = Detail["MaxTemp"]['max'] - Detail["MinTemp"]['min']
Detail = Detail.loc[:,["MaxTemp","MinTemp","Diff of Temp","Rainfall"]]  # lowest and Highest temperature in years
Detail

MaxTemp               MinTemp       Diff of Temp  \
                          max       mean       mean   min                
Location         Year                                                    
Adelaide         2008    36.6  20.669399  10.843716   0.7         35.9   
                 2009    45.7  23.442740  13.144932   2.9         42.8   
                 2010    42.8  22.443562  12.582740   2.2         40.6   
                 2011    42.9  22.677910  12.882687   1.3         41.6   
                 2012    41.6  22.247164  12.003881   2.2         39.4   
                 2013    45.0  22.867359  12.719585   3.6         41.4   
                 2014    45.1  23.533699  12.835890   0.9         44.2   
                 2015    44.1  23.163462  12.448352   1.8         42.3   
                 2016    41.3  22.863661  12.346721   2.0         39.3   
                 2017    42.4  24.977143  13.678977   1.7         40.7   
Albany           2008    27.7  20.709677  14.486667  10.5         17.2   
                 2009    36.4  19.554521  12.606648   6.0         30.4   
                 2010    38.9  19.820000  12.713151   5.5         33.4   
                 2011    37.9  20.204776  12.984179   1.8         36.1   
                 2012    37.3  20.642643  12.721148   4.5         32.8   
                 2013    38.1  19.985373  13.011343   5.2         32.9   
                 2014    32.8  20.256111  13.219274   4.5         28.3   
                 2015    37.5  20.269274  13.189972   5.0         32.5   
                 2016    39.0  19.432749  12.535014   4.8         34.2   
                 2017    35.9  20.901235  14.107534   7.8         28.1   
Albury           2008    33.9  27.293548  13.509677   7.4         26.5   
                 2009    44.8  23.241209   9.652198  -2.0         46.8   
                 2010    42.4  21.812465   9.627424  -2.0         44.4   
                 2011    39.8  21.785542   9.385843  -1.6         41.4   
                 2012    38.9  21.168358   8.011940  -2.5         41.4   
                 2013    43.4  22.417262   8.769345  -1.3         44.7   
                 2014    43.6  23.364463   9.828650  -2.8         46.4   
                 2015    41.5  22.779726   9.557534  -2.4         43.9   
                 2016    43.0  22.734153  10.234973  -1.3         44.3   
                 2017    43.7  25.147727  10.861364  -1.1         44.8   
AliceSprings     2008    40.9  34.641935  20.903226  13.1         27.8   
                 2009    43.2  29.954247  14.299452   0.2         43.0   
                 2010    41.3  26.624658  12.725479  -2.7         44.0   
                 2011    43.0  28.565672  12.584478  -3.4         46.4   
                 2012    42.7  28.833433  10.989254  -5.2         47.9   
                 2013    44.4  30.365282  13.471513  -1.1         45.5   
                 2014    44.9  29.701370  13.240934  -3.0         47.9   
                 2015    42.5  29.135714  12.675616  -5.0         47.5   
                 2016    42.6  29.398087  13.807104  -4.0         46.6   
                 2017    41.9  31.220571  14.150568  -2.9         44.8   
BadgerysCreek    2009    42.9  24.468767  10.950418  -1.3         44.2   
                 2010    43.0  23.432133  11.256389  -3.0         46.0   
                 2011    41.5  23.452836  10.933433  -1.7         43.2   
                 2012    38.4  22.931940   9.821194  -1.1         39.5   
                 2013    45.8  24.295238  10.605045  -0.4         46.2   
                 2014    41.6  24.360221  11.580939  -1.5         43.1   
                 2015    41.8  23.635912  11.343767  -1.3         43.1   
                 2016    40.8  24.684483  11.147989  -1.9         42.7   
                 2017    46.4  25.886364  13.735795  -0.1         46.5   
Ballarat         2008    32.2  21.377419   8.770968   2.9         29.3   
                 2009    44.1  18.831781   7.252329  -2.4         46.5   
                 2010    3

In [93]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 25 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   Date           145460 non-null  datetime64[ns]
 1   Location       145460 non-null  category      
 2   MinTemp        143975 non-null  float64       
 3   MaxTemp        144199 non-null  float64       
 4   Rainfall       142199 non-null  float64       
 5   Evaporation    82670 non-null   float64       
 6   Sunshine       75625 non-null   float64       
 7   WindGustDir    135134 non-null  category      
 8   WindGustSpeed  135197 non-null  float64       
 9   WindDir9am     134894 non-null  category      
 10  WindDir3pm     141232 non-null  category      
 11  WindSpeed9am   143693 non-null  float64       
 12  WindSpeed3pm   142398 non-null  float64       
 13  Humidity9am    142806 non-null  float64       
 14  Humidity3pm    140953 non-null  float64       
 15  

In [94]:
df.drop(labels=["Sunshine","Evaporation","Cloud9am","Cloud3pm"],axis=1,inplace= True) # Dropped N

In [95]:
df.head(5)

Date Location  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed  \
0 2008-12-01   Albury     13.4     22.9       0.6           W           44.0   
1 2008-12-02   Albury      7.4     25.1       0.0         WNW           44.0   
2 2008-12-03   Albury     12.9     25.7       0.0         WSW           46.0   
3 2008-12-04   Albury      9.2     28.0       0.0          NE           24.0   
4 2008-12-05   Albury     17.5     32.3       1.0           W           41.0   

  WindDir9am WindDir3pm  WindSpeed9am  ...  Humidity9am  Humidity3pm  \
0          W        WNW          20.0  ...         71.0         22.0   
1        NNW        WSW           4.0  ...         44.0         25.0   
2          W        WSW          19.0  ...         38.0         30.0   
3         SE          E          11.0  ...         45.0         16.0   
4        ENE         NW           7.0  ...         82.0         33.0   

   Pressure9am  Pressure3pm  Temp9am  Temp3pm  RainToday RainTomorrow  Year  \
0       1007.7       1007.1     16.9     21.8         No           No  2008   
1       1010.6       1007.8     17.2     24.3         No           No  2008   
2       1007.6       1008.7     21.0     23.2         No           No  2008   
3       1017.6       1012.8     18.1     26.5         No           No  2008   
4       1010.8       1006.0     17.8     29.7         No           No  2008   

      Month  
0  December  
1  December  
2  December  
3  December  
4  December  

[5 rows x 21 columns]

In [96]:
df.nunique()

Date             3436
Location           49
MinTemp           389
MaxTemp           505
Rainfall          681
WindGustDir        16
WindGustSpeed      67
WindDir9am         16
WindDir3pm         16
WindSpeed9am       43
WindSpeed3pm       44
Humidity9am       101
Humidity3pm       101
Pressure9am       546
Pressure3pm       549
Temp9am           441
Temp3pm           502
RainToday           2
RainTomorrow        2
Year               11
Month              12
dtype: int64

In [97]:
df.dtypes

Date             datetime64[ns]
Location               category
MinTemp                 float64
MaxTemp                 float64
Rainfall                float64
WindGustDir            category
WindGustSpeed           float64
WindDir9am             category
WindDir3pm             category
WindSpeed9am            float64
WindSpeed3pm            float64
Humidity9am             float64
Humidity3pm             float64
Pressure9am             float64
Pressure3pm             float64
Temp9am                 float64
Temp3pm                 float64
RainToday              category
RainTomorrow           category
Year                      int64
Month                  category
dtype: object

In [98]:
# Converting "RainTOday" and "Rain Tomorrow" into numerical 

#  Drop rows with null

In [99]:
df.dropna(inplace = True)

In [100]:
df["RainToday"].isnull().sum()

0

In [101]:
df["RainToday"].unique()  # categorical data check

['No', 'Yes']
Categories (2, object): ['No', 'Yes']

In [102]:
def yesno(rain):
    if rain == "Yes":
        return 1
    else:
        return 0

In [103]:
df["RainToday"] = df["RainToday"].apply(yesno)   # converted Yes with 1 and No by 0
df["RainTomorrow"] = df["RainTomorrow"].apply(yesno)

In [104]:
df.head(3)

Date Location  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed  \
0 2008-12-01   Albury     13.4     22.9       0.6           W           44.0   
1 2008-12-02   Albury      7.4     25.1       0.0         WNW           44.0   
2 2008-12-03   Albury     12.9     25.7       0.0         WSW           46.0   

  WindDir9am WindDir3pm  WindSpeed9am  ...  Humidity9am  Humidity3pm  \
0          W        WNW          20.0  ...         71.0         22.0   
1        NNW        WSW           4.0  ...         44.0         25.0   
2          W        WSW          19.0  ...         38.0         30.0   

   Pressure9am  Pressure3pm  Temp9am  Temp3pm  RainToday RainTomorrow  Year  \
0       1007.7       1007.1     16.9     21.8          0            0  2008   
1       1010.6       1007.8     17.2     24.3          0            0  2008   
2       1007.6       1008.7     21.0     23.2          0            0  2008   

      Month  
0  December  
1  December  
2  December  

[3 rows x 21 columns]

In [105]:
df.nunique()

Date             3417
Location           44
MinTemp           377
MaxTemp           413
Rainfall          636
WindGustDir        16
WindGustSpeed      66
WindDir9am         16
WindDir3pm         16
WindSpeed9am       39
WindSpeed3pm       41
Humidity9am       101
Humidity3pm       101
Pressure9am       542
Pressure3pm       541
Temp9am           397
Temp3pm           415
RainToday           2
RainTomorrow        2
Year               11
Month              12
dtype: int64

In [106]:
df.dtypes

Date             datetime64[ns]
Location               category
MinTemp                 float64
MaxTemp                 float64
Rainfall                float64
WindGustDir            category
WindGustSpeed           float64
WindDir9am             category
WindDir3pm             category
WindSpeed9am            float64
WindSpeed3pm            float64
Humidity9am             float64
Humidity3pm             float64
Pressure9am             float64
Pressure3pm             float64
Temp9am                 float64
Temp3pm                 float64
RainToday              category
RainTomorrow           category
Year                      int64
Month                  category
dtype: object

# converting Location,WindGustDir, WindDir9am, WindDir3pm in Binary Encoding

In [108]:
pip install category_encoders

Note: you may need to restart the kernel to use updated packages.


In [109]:
from category_encoders import BinaryEncoder
encoder = BinaryEncoder(cols = ["Location","WindGustDir","WindDir9am","WindDir3pm"])
newdata = encoder.fit_transform(df[["Location","WindGustDir","WindDir9am","WindDir3pm"]])
df = pd.concat([df,newdata],axis=1)
df.drop(["Location""WindGustDir","WindDir9am","WindDir3pm"], axis=1, inplace = True)


Date Location  MinTemp  MaxTemp  Rainfall  WindGustSpeed  \
0 2008-12-01   Albury     13.4     22.9       0.6           44.0   
1 2008-12-02   Albury      7.4     25.1       0.0           44.0   
2 2008-12-03   Albury     12.9     25.7       0.0           46.0   
3 2008-12-04   Albury      9.2     28.0       0.0           24.0   
4 2008-12-05   Albury     17.5     32.3       1.0           41.0   
5 2008-12-06   Albury     14.6     29.7       0.2           56.0   
6 2008-12-07   Albury     14.3     25.0       0.0           50.0   
7 2008-12-08   Albury      7.7     26.7       0.0           35.0   
8 2008-12-09   Albury      9.7     31.9       0.0           80.0   
9 2008-12-10   Albury     13.1     30.1       1.4           28.0   

   WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  ...  WindDir9am_0  \
0          20.0          24.0         71.0         22.0  ...             0   
1           4.0          22.0         44.0         25.0  ...             0   
2          19.0          26.0         38.0         30.0  ...             0   
3          11.0           9.0         45.0         16.0  ...             0   
4           7.0          20.0         82.0         33.0  ...             0   
5          19.0          24.0         55.0         23.0  ...             0   
6          20.0          24.0         49.0         19.0  ...             0   
7           6.0          17.0         48.0         19.0  ...             0   
8           7.0          28.0         42.0          9.0  ...             0   
9          15.0          11.0         58.0         27.0  ...             0   

   WindDir9am_1  WindDir9am_2  WindDir9am_3 WindDir9am_4 WindDir3pm_0  \
0             0             0             0            1            0   
1             0             0             1            0            0   
2             0             0             0            1            0   
3             0             0             1            1            0   
4             0             1             0            0            0   
5             0             0             0            1            0   
6             0             1             0            1            0   
7             0             1             1            0            0   
8             0             0             1            1            0   
9             0             1             1            1            0   

   WindDir3pm_1 WindDir3pm_2  WindDir3pm_3  WindDir3pm_4  
0             0            0             0             1  
1             0            0             1             0  
2             0            0             1             0  
3             0            0             1             1  
4             0            1             0             0  
5             0            1             0             1  
6             0            1             0             1  
7             0            1             0             1  
8             0            1             0             0  
9             0            1             1             0  

[10 rows x 33 columns]